File: hook-epic/src/main/java/io/github/xanderwang/hook/epic/MethodEpicParam.java
package io.github.xanderwang.hook.epic;

import io.github.xanderwang.hook.core.MethodParam;

import java.lang.reflect.Member;

import de.robv.android.xposed.XC_MethodHook.MethodHookParam;

public class MethodEpicParam implements MethodParam {

  MethodHookParam methodHookParam;

  public MethodEpicParam() {
  }

  public void setMethodHookParam(MethodHookParam methodHookParam) {
    this.methodHookParam = methodHookParam;
  }

  @Override
  public Object getResult() {
    return methodHookParam.getResult();
  }

  @Override
  public void setResult(Object result) {
    methodHookParam.setResult(result);
  }

  @Override
  public Throwable getThrowable() {
    return methodHookParam.getThrowable();
  }

  @Override
  public boolean hasThrowable() {
    return methodHookParam.hasThrowable();
  }

  @Override
  public void setThrowable(Throwable throwable) {
    methodHookParam.setThrowable(throwable);
  }

  @Override
  public Object getResultOrThrowable() throws Throwable {
    return methodHookParam.getResultOrThrowable();
  }

  @Override
  public Object[] getArgs() {
    return methodHookParam.args;
  }

  @Override
  public Object getThisObject() {
    return methodHookParam.thisObject;
  }

  @Override
  public Member getMethod() {
    return methodHookParam.method;
  }
}

File: hook-epic/src/main/java/io/github/xanderwang/hook/epic/MethodEpicParam.kt
package io.github.xanderwang.hook.epic

import de.robv.android.xposed.XC_MethodHook
import io.github.xanderwang.hook.core.MethodParam
import java.lang.reflect.Member

class MethodEpicParam : MethodParam {

    var methodHookParam: XC_MethodHook.MethodHookParam? = null

    override var result: Any?
        get() = methodHookParam?.result
        set(result) {
            methodHookParam?.result = result
        }

    override var throwable: Throwable?
        get() = methodHookParam?.throwable
        set(throwable) {
            methodHookParam?.throwable = throwable
        }

    override fun hasThrowable(): Boolean {
        return methodHookParam?.hasThrowable() == true
    }

    @get:Throws(Throwable::class)
    override val resultOrThrowable: Any?
        get() = methodHookParam?.resultOrThrowable


    override var args: Array<Any?>?
        get() = methodHookParam?.args
        set(args) {
            methodHookParam?.args = args
        }

    override var thisObject: Any?
        get() = methodHookParam?.thisObject
        set(thisObject) {
            methodHookParam?.thisObject = thisObject
        }

    override var method: Member?
        get() = methodHookParam?.method
        set(method) {
            methodHookParam?.method = method
        }
}